kube-proxy IPVS: Fix IPVS availability check

vfreex opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
The current implementation of `CanUseIPVSProxier()` reads `/proc/modules`
to check whether IPVS related kernel modules can be loaded.
  
You might get a false-negative when the kernel modules are installed to
the system but haven't been loaded into the kernel.

This patch firstly try to run `modprobe` to load specified kernel
modules, then just log warnings if error occured.
Secondly, it will check loaded kernel modules by reading
`/proc/modules`, return an error if any required module is missing.

This change will not break the compatability of existing implementation.
Running kube-proxy in a container without mounting `/lib/modules` will
cause `modprobe` warnings, but not raise an error if all required modules are
present in the host kernel.

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
 Fix IPVS availability check
```

shiywang commented:
/ok-to-test

vfreex commented:
/retest

shiywang commented:
@m1093782566 @kubernetes/huawei ptal

m1093782566 commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -614,24 +614,15 @@ func (em proxyEndpointsMap) unmerge(other proxyEndpointsMap) {
-	if len(modules) != 0 {
-		return false, fmt.Errorf("Failed to load kernel modules: %v", modules)
+	for _, kmod := range ipvsModules {
+		err := utilexec.New().Command("modprobe", "--dry-run", kmod).Run()
```
1. Currently, kube-proxy is running in container - of course, kube-proxy can run outside of container. I think the output of `modprobe` exec inside container is empty?

2. Did you installed `modprobe` command? Sorry, I didn't see it.

3. I admit better documentation and comments are needed. @Lion-Wei Can you do that?

vfreex replied:
@m1093782566 
Thanks for you comment.
1. By default kernel modules are not included in containers, so neither auto-load mechanism nor `modprobe` doesn't work out of the box.  A workaround is to mount the kube-proxy container with ` -v /lib/modules:/host/lib/modules:ro` (of cause, privileged mode is needed). See [discussions at StackOverflow](https://stackoverflow.com/questions/33013539/docker-loading-kernel-modules).
2. `modprobe` should be builtin in most Linux distributions, but I haven't checked if the kube-proxy image contains it.

vfreex replied:
I admit that I am not sure which approach is more appropriate at the moment. Looking forward to the opinions from your sides.

vfreex replied:
Additional info added below.

vfreex commented:
I did some investigations after submitting this PR, here are my findings:

Precondition: A node running Linux with IPVS kernel modules installed.

1. When kube-proxy is running outside the container, the IPVS kernel modules will be loaded automatically when the kernel first needs it. Before that happens, `/proc/modules` does not contain them.

2. When kube-proxy is running inside the container, the auto-load mechanism will only work when mounting the kube-proxy with `/lib/modules:/lib/modules:ro` (I've tested that). Considering kube-proxy is already a privileged container, it should not increase the security risk.

3. Actually, https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/docker/libnetwork/ipvs/netlink.go#L60 already has logic to determine the availability of IPVS. It will try run `modprobe` first then call `getIPVSFamily()`. By following the calling chain, we can see this function will be called when `ipvs.New()` is called in the first time. Although it may not return an error immediately if the check fails.





vendor/github.com/docker/libnetwork/ipvs/netlink.go:60

vfreex commented:
Any suggestions?

Lion-Wei commented:
@vfreex @m1093782566 , I don't think it can load any kernel modules automatically. In current version, if kube-proxy is running outside the container, we must use `modprobe` to load kernel modules manually. I will write a document about that as soon as possible. 
I think first we should solve is how to load kernel modules automatically, especially when running inside the container. That may need more advise from @m1093782566 .

vfreex commented:
Hi, @Lion-Wei, thanks for your comment.
Kernel modules loading should be handled automatically by [udev](https://www.kernel.org/pub/linux/utils/kernel/hotplug/udev/udev.html).

You mentioned that `kube-proxy is running outside the container, we must use modprobe to load kernel modules manually`. This is true, but it is not related to the auto-loading.
This is because `CanUseIPVSProxier()` is not working properly, which is what this PR is going to fix.

vfreex commented:
/sig network

m1093782566 commented:
@vfreex 

> In other words, even if the required kernel modules are installed, when there are no any virtual services in the system, the kernel may not load those kernel modules.

I am afraid I didn't fully understand it. Here is my test:

Ubuntu 14.04 GNU/Linux 3.13.0-24-generic x86_64


```
root@SZX1000353065:~# lsmod | grep ip_vs

root@SZX1000353065:~# ipvsadm -ln
The program 'ipvsadm' is currently not installed. You can install it by typing:
apt-get install ipvsadm

root@SZX1000353065:~# modprobe ip_vs

root@SZX1000353065:~# cat /proc/modules | grep ip_vs
ip_vs 119517 0 - Live 0xffffffffa0360000
nf_conntrack 96976 1 ip_vs, Live 0xffffffffa037f000
libcrc32c 12644 3 ip_vs,btrfs,xfs, Live 0xffffffffa00eb000
```

I did't create any ipvs service, but after `modprobe ip_vs`, I find `ip_vs` existing in `/proc/modules`?

m1093782566 commented:
> When kube-proxy is running inside the container, the auto-load mechanism will only work when mounting the kube-proxy with /lib/modules:/lib/modules:ro (I've tested that). Considering kube-proxy is already a privileged container, it should not increase the security risk.

Personally, I like this idea if it can works in every OS distribution and architecture - we may need more test. I think @dhilipkumars has already done some interesting tests, can you give us your opinions :)

vfreex commented:
@m1093782566 
Yes, `modprobe ip_vs` absolutely works if the module .ko files are residing in `/lib/modules/<kernel version>`.
My point is, in most cases, `modprobe <some module>` is not necessary because the kernel can load the modules by need. It would be painful for a large cluster because every time a node boots, you need to run `modprobe` manually. It would be much convenient if this can be done in an automated manner. 

I am in favor of your suggestion that we may need more tests.

vfreex commented:
A similar demo:
```
~% lsmod | grep ip_vs     
~% ipvsadm
zsh: ipvsadm: command not found...
% sudo dnf install ipvsadm
% lsmod | grep ip_vs   
% sudo ipvsadm -A -t 127.0.0.1:8080 -s rr
% lsmod | grep ip_vs
ip_vs_rr               16384  1
ip_vs                 151552  3 ip_vs_rr
nf_conntrack          126976  12 ip_vs,nf_conntrack_ipv6,nf_conntrack_ipv4,ipt_MASQUERADE,nf_conntrack_broadcast,nf_conntrack_netbios_ns,xt_CT,nf_nat_ipv6,nf_nat_masquerade_ipv4,xt_conntrack,nf_nat_ipv4,nf_nat
libcrc32c              16384  3 ip_vs,nf_conntrack,nf_nat

```

m1093782566 commented:
@vfreex 

> Installing the ipvsadm package doesn't mean the kernel modules will be loaded. The auto-load happens when the kernel actually needs it. For example, after executing an ipvsadm command.

ACK.

I mean, when there is no ipvs service in my machine, but after I loading the kernel modules by hand outside the container, I find these modules exist in `/proc/modules` of container. So, I think you can't say the way of checking `/proc/modules` is wrong.

The question is how to automatically load kernel modules(inside container or outside container).

vfreex commented:
@m1093782566 
> I mean, when there is no ipvs service in my machine, but after I loading the kernel modules by hand outside the container, I find these modules exist in /proc/modules of container. So, I think you can't say the way of checking /proc/modules is wrong.
Yeah, because the container shares the same kernel with the machine.
Sure, probably it is not appropriate to say that checking /proc/modules is wrong. It just has a false negative condition. In other words, when the function returns true, the result is 100% correct. But when it returns false, it may be correct (kernel version is too old), but there is a possibility that the kernel isn't using it, not IPVS is not supported.

> The question is how to automatically load kernel modules(inside container or outside container).
Sure, they are the same thing in different aspects.

vfreex commented:
/retest

dhilipkumars reviewed (commented):
IMHO Highly recommend way would be to de-couple and  insert such pre-start tests in the initContainers in kube-proxy's daemonset definition. 

We should also document this elaborately in the user guides on how users can safely check ipvs prerequisite before starting to use them in their k8s setup?

vfreex commented:
The latest change reverts the old logic, which is reading `/proc/modules` for loaded kernel modules, and only run `modprobe --dry-run` check for the missing ones.
This is for better performance and comparability.

m1093782566 commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -629,9 +629,14 @@ func CanUseIPVSProxier() (bool, error) {
+	for kmod := range modules {
+		err := utilexec.New().Command("modprobe", "--dry-run", kmod).Run()
+		if err != nil {
+			return false, fmt.Errorf("Failed to load kernel module: %v", kmod)
```
I think it should not return error directly here since it's a "best effort" way. Log warning instead?

m1093782566 commented:
self /assign @m1093782566

m1093782566 commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -629,9 +629,14 @@ func CanUseIPVSProxier() (bool, error) {
+
+	// Now check if the missing kernel modules are there to be loaded
+	for kmod := range modules {
+		err := utilexec.New().Command("modprobe", "--dry-run", kmod).Run()
```
Please check my test:

```
# cat /proc/modules | grep ip_vs
ip_vs_dh 12688 0 - Live 0xffffffffa05e5000
ip_vs_lc 12516 0 - Live 0xffffffffa05d5000
ip_vs_sh 12688 0 - Live 0xffffffffa05c5000
ip_vs_wrr 12697 0 - Live 0xffffffffa05c0000
ip_vs_rr 12600 4 - Live 0xffffffffa05bb000
ip_vs 141092 14 ip_vs_dh,ip_vs_lc,ip_vs_sh,ip_vs_wrr,ip_vs_rr, Live 0xffffffffa0592000
nf_conntrack 111302 7 nf_conntrack_netlink,ip_vs,nf_nat_masquerade_ipv4,nf_nat_ipv4,nf_nat,nf_conntrack_ipv4,xt_conntrack, Live 0xffffffffa045a000
libcrc32c 12644 2 ip_vs,dm_persistent_data, Live 0xffffffffa0349000

# modprobe --dry-run ip_vs
modprobe: ERROR: ../libkmod/libkmod.c:557 kmod_search_moddep() could not open moddep file '/lib/modules/3.10.0-514.26.2.el7.x86_64/modules.dep.bin'
```

The updated logic doesn't hit my expectation.

m1093782566 replied:
I would suggest:

Firstly, modprobe ip_vs_* modules, then just log warnings if error occured.

Secondly,  check fields in `/proc/modules`, then return error if required filed missing.

vfreex replied:
Have you mounted /lib/modules from the node to container?
`check fields in /proc/modules, then return error if required filed missing`
This approach makes sense to me.

m1093782566 replied:
> Have you mounted /lib/modules from the node to container?

NO.

m1093782566 commented:
/assign @m1093782566

m1093782566 commented:
/retest

m1093782566 commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -614,10 +614,18 @@ func (em proxyEndpointsMap) unmerge(other proxyEndpointsMap) {
+	for _, kmod := range ipvsModules {
+		err := utilexec.New().Command("modprobe", "--", kmod).Run()
+		if err != nil {
+			glog.Warningf("Failed to load kernel module: %v", kmod)
```
nit:

The log message is identical with https://github.com/vfreex/kubernetes/blob/5a0588c34d13c77166d584c61aa59d1621ddb191/pkg/proxy/ipvs/proxier.go#L642

I would prefer a verbose message here, e.g. "Failed to load kernel module: %v using modprobe, please ignore it when kube-proxy is running inside container without mounting /lib/modules". WDYT ^-^

BTW, please fix the log message https://github.com/vfreex/kubernetes/blob/5a0588c34d13c77166d584c61aa59d1621ddb191/pkg/proxy/ipvs/proxier.go#L642, what about "Required kernel module %s is not loaded"?

vfreex committed:
kube-proxy IPVS: Fix IPVS availability check

The current implementation of `CanUseIPVSProxier()` reads `/proc/modules`
to check whether IPVS related kernel modules can be loaded.

You might get a false-negative when the kernel modules are installed to
the system but haven't been loaded into the kernel.

This patch firstly try to run `modprobe` to load specified kernel
modules, then just log warnings if error occured.
Secondly, it will check loaded kernel modules by reading
`/proc/modules`, return an error if any required module is missing.

This change will not break the compatability of existing implementation.
Running kube-proxy in a container without mounting `/lib/modules` will
cause `modprobe` warnings, but not raise an error if all required modules are
present in the host kernel.

m1093782566 commented:
/lgtm from my side.

/assign @thockin for phase-2 review :)

m1093782566 commented:
/assign @thockin

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -614,10 +614,19 @@ func (em proxyEndpointsMap) unmerge(other proxyEndpointsMap) {
 func CanUseIPVSProxier() (bool, error) {
+	// Try to load IPVS required kernel modules using modprobe
+	for _, kmod := range ipvsModules {
+		err := utilexec.New().Command("modprobe", "--", kmod).Run()
```
PR comment says dry-run, this does not ?

thockin replied:
Does this work if kube-proxy is in a container?

@MrHohn

MrHohn replied:
I believe this depends on the works on https://github.com/kubernetes/kubernetes/pull/52003.

m1093782566 replied:
> PR comment says dry-run, this does not ?

I did a simple test and found `modprobe --dry-run ${kmod}` == `modprobe -- ${kmod}`

@vfreex right?

m1093782566 replied:
> Does this work if kube-proxy is in a container?

this updated logic does no harm when kube-proxy is in a container.

thockin replied:
How is `--` the same as `--dry-run` ?  That's not in the docs and doesn't make much sense.

vfreex replied:
@thockin Sorry I removed the `--dry-run` option in latter commits.
The latest logic is, try to running `modprobe` to load required kernel modules, and ignore the error if `modprobe` fails. Then check if all those modules are already loaded in the kernel.
Successful running `modprobe` in a container requires `/lib/moudules` on node is mounted to the container. If not doing so, `modprobe` will fail but the error will be ignored.

vfreex replied:
`--` is different from `--dry-run`. `--dry-run` only checks the existence of the modules without actually loading it.
`--` is only an explicit separator to prevent from parsing options: `modprobe -- ${kmod}` is the same as `modprobe ${kmod}`.

vfreex commented:
Rebased and amended the description of the PR.

m1093782566 commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -648,10 +648,19 @@ func (em proxyEndpointsMap) unmerge(other proxyEndpointsMap) {
 func CanUseIPVSProxier() (bool, error) {
+	// Try to load IPVS required kernel modules using modprobe
+	for _, kmod := range ipvsModules {
+		err := utilexec.New().Command("modprobe", "--", kmod).Run()
```
Could you please add some explanation on `--`?

vfreex replied:
It is not required. For example, if you want to remove a file called `-f`, you cannot delete it with `rm -f` because `-f` will be treated as an option not a filename. You have to do it with `rm -- -f`.
The double dash is not something much here, just my personal habit.
https://unix.stackexchange.com/questions/11376/what-does-double-dash-mean-also-known-as-bare-double-dash.

m1093782566 commented:
/lgtm

brendandburns commented:
/retest

vfreex commented:
/retest

vfreex commented:
/retest

vfreex commented:
/retest

vfreex commented:
/retest pull-kubernetes-e2e-gce-bazel

vfreex commented:
/test pull-kubernetes-e2e-gce-bazel

vfreex commented:
/test all

vfreex commented:
/retest

On Thu, Oct 26, 2017, 7:01 PM k8s-ci-robot <notifications@github.com> wrote:

> @vfreex <https://github.com/vfreex>: The following tests *failed*, say
> /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-e2e-gce-bazel 75617b4
> <https://github.com/kubernetes/kubernetes/commit/75617b4c3948764010b9f6dcb307902554e1fa43>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51874/pull-kubernetes-e2e-gce-bazel/35459/> /test
> pull-kubernetes-e2e-gce-bazel
> pull-kubernetes-unit 75617b4
> <https://github.com/kubernetes/kubernetes/commit/75617b4c3948764010b9f6dcb307902554e1fa43>
> link
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51874/pull-kubernetes-unit/60765/> /test
> pull-kubernetes-unit
>
> Full PR test history <https://k8s-gubernator.appspot.com/pr/51874>. Your
> PR dashboard <https://k8s-gubernator.appspot.com/pr/vfreex>. Please help
> us cut down on flakes by linking to
> <https://github.com/kubernetes/community/blob/master/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md>.
> If you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://github.com/kubernetes/test-infra/blob/master/commands.md>.
>
> —
>
>
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/51874#issuecomment-339629775>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJezDSlN7GBl_tckkz7ePwW3c-sxpMOWks5swGaSgaJpZM4PLOJz>
> .
>

m1093782566 commented:
@thockin 

Since #52003 alredy goes in, I guess we can let this PR in as well?

Lion-Wei commented:
/retest

Lion-Wei commented:
Our ipvs test CI job do need this pr to load kernel mode, now that job is running periodically, so we do need this pr be approved ASAP. @thockin @brendandburns @m1093782566

vfreex commented:
flake https://github.com/kubernetes/kubernetes/issues/54873

cmluciano commented:
/retest

m1093782566 commented:
The change still looks good to me

/lgtm

m1093782566 commented:
ping @thockin for approval.

Lion-Wei commented:
/test pull-kubernetes-e2e-kops-aws

thockin commented:
/approve no-issue
